File: leetcode/Python/largest-divisible-subset.py
class Solution:
    def largestDivisibleSubset(self, nums: List[int]) -> List[int]:
        dp, m_idx, idx = [list() for _ in range(len(nums))], 0, -1
        nums.sort()
        for i in range(len(nums)):
            idx = -1
            if i==0:
                dp[i].append(nums[i])
                continue
            for j in range(i):
                if nums[i]%dp[j][-1]==0 and (idx==-1 or len(dp[idx])<len(dp[j])):
                    idx = j
            if idx>=0:
                dp[i] = dp[idx][:]
            dp[i].append(nums[i])
            if len(dp[m_idx])<len(dp[i]):
                m_idx=i
        return dp[m_idx]

File: leetcode/C++/largest-divisible-subset.cpp
class Solution {
public:
    vector<int> largestDivisibleSubset(vector<int>& nums) {
        sort(nums.begin(),nums.end());
        vector<vector<int>> dp(nums.size());
        int m_idx=0;
        for(int i=0,idx=-1; i<nums.size(); ++i){
            idx=-1;
            if(i==0) {dp[i].push_back(nums[i]); continue;}
            for(int j=0; j<i; ++j){
                if(nums[i]%dp[j].back()==0 && (idx==-1 || dp[idx].size()<dp[j].size())) idx=j;
            }
            if(idx>=0) dp[i]=dp[idx];
            dp[i].push_back(nums[i]);
            if(dp[m_idx].size()<dp[i].size()) m_idx=i;
        }
        return dp[m_idx];
    }
};